README.md
# Open WebView KSU

![Open WebView](https://raw.githubusercontent.com/FoxeiZ/open-webview-ksu/main/img/logo.png)

This script helps you create the module to replace your system webview though KernelSU.

>**ATTENTION!** There is a bug that still needs to be fixed where sometimes you may find that the installed webview doesn't work. To fix this bug you need to manually install the webview, to do this just:<br/><br/>- go to fdroid and re-download the latest version<br/>- manually download the apk and install it

## Compatibility

- Android 8+
- KernelSU v0.6.1+ (11041)

## Tested Device

- [LOS 20](https://lineageos.org/)
- [Corvus](https://github.com/Corvus-AOSP)
- And more...

## DISCLAIMER

Before flash module, please read below:
>I AM NOT RESPONSIBLE SOME YOUR FEATURE FROM YOUR DEVICE IF DOESN'T WORK PROPERLY. BEFORE FLASH THE MODULE, PLEASE READ LINE CODES AND SELECT GOOGLE SERVICES YOU NEEDS. YOU ARE FLASHING THIS MAGISK MODULE AND ITS YOUR CHOICE TO DO IT OR NOT TO DO IT AND YOU'RE THE ONE DOING IT. I JUST WANT TO HELP OTHERS OUT.

## Build step

```
# Run this on any terminal with root access

/data/adb/ksu/bin/busybox wget -qO /data/local/tmp/build.sh https://github.com/FoxeiZ/open-webview-ksu/raw/main/build.sh
ASH_STANDALONE=1 /data/adb/ksu/bin/busybox sh /data/local/tmp/build.sh

```

## Support

If you found this helpful, please consider supporting development with a [coffe](https://www.paypal.me/f3ff0). Alternatively, you can contribute to the project by reporting bugs and doing PR. All support is appreciated!

## Features

- Works on any device running Android 8.0+ and Magisk 20.4+
- Replace the webview with one of:
    1. [Bromite WebView](https://github.com/bromite/bromite)
    2. [Mulch](https://gitlab.com/divested-mobile/mulch)

## Credits

- Original [repo](https://github.com/Magisk-Modules-Alt-Repo/open_webview) by [F3FFO](https://github.com/F3FFO/)
- [aapt-binary](https://github.com/JonForShort/android-tools)
- [MMT-Extended-Next](https://github.com/symbuzzer/MMT-Extended-Next) by [symbuzzer](https://github.com/symbuzzer)
- [Bromite](https://github.com/bromite/bromite)
- [DivestOS](https://gitlab.com/divested-mobile)
- [cUrl](https://github.com/curl/curl)
- [cUrl binary](https://github.com/F3FFO/compile_zlib_openssl_curl_android)
- [Zipsigner and zip](https://github.com/Magisk-Modules-Repo/zipsigner) by [osm0sis](https://github.com/osm0sis)

## License

Copyright 2023 F3FFO

The source code is available under [GPL-3.0](https://github.com/Magisk-Modules-Alt-Repo/open_fonts/blob/master/LICENSE)

## Change logs

# v2.1.0_1

- Porting to ksu
- Change aapt binary
- Reduce zip size
- Unpack all overlays zip file

# v2.1.0

- Add Mulch webview
- Add status in the module description
- Update bromite to v108.0.5359.156
- Bug fix

# v2.0.0

- Reworked the installation logic
- Add cleaning logic for dalvik cache
- Update bromite to v106.0.5249.163
- Bug fix

# v1.1.0

- Add curl binary
- Bug fix

# v1.0.1

- Bug fix

# v1.0.0

- Initial release


CHANGELOG.md
**If you found this helpful, please consider supporting development with a [coffe](https://www.paypal.me/f3ff0). Alternatively, you can contribute to the project by reporting bugs and doing PR. All support is appreciated!**

# v2.1.0_1

- Porting to ksu
- Change aapt binary
- Reduce zip size
- Unpack all overlays zip file

# v2.1.0

- Add Mulch webview
- Add status in the module description
- Update bromite to v108.0.5359.156
- Bug fix

# v2.0.0

- Reworked the installation logic
- Add cleaning logic for dalvik cache
- Update bromite to v106.0.5249.163
- Bug fix

# v1.1.0

- Add curl binary
- Bug fix

# v1.0.1

- Bug fix

# v1.0.0

- Initial release


module/customize.sh
##########################################################################################
#
# MMT Extended Config Script
#
##########################################################################################

##########################################################################################
# Config Flags
##########################################################################################

# Uncomment and change 'MINAPI' and 'MAXAPI' to the minimum and maximum android version for your mod
# Uncomment DYNLIB if you want libs installed to vendor for oreo+ and system for anything older
# Uncomment DEBUG if you want full debug logs (saved to /sdcard)
MINAPI=26
#MAXAPI=25
#DYNLIB=true
DEBUG=true

##########################################################################################
# Replace list
##########################################################################################

# List all directories you want to directly replace in the system
# Check the documentations for more info why you would need this

# Construct your list in the following format
# This is an example
REPLACE_EXAMPLE="
/system/app/Youtube
/system/priv-app/SystemUI
/system/priv-app/Settings
/system/framework
"

# Construct your own list here
REPLACE="
"

##########################################################################################
# Permissions
##########################################################################################

set_permissions() {
  : # Remove this if adding to this function

  # Note that all files/folders in magisk module directory have the $MODPATH prefix - keep this prefix on all of your files/folders
  # Some examples:
  
  # For directories (includes files in them):
  # set_perm_recursive  <dirname>                <owner> <group> <dirpermission> <filepermission> <contexts> (default: u:object_r:system_file:s0)
  
  # set_perm_recursive $MODPATH/system/lib 0 0 0755 0644
  # set_perm_recursive $MODPATH/system/vendor/lib/soundfx 0 0 0755 0644

  # For files (not in directories taken care of above)
  # set_perm  <filename>                         <owner> <group> <permission> <contexts> (default: u:object_r:system_file:s0)
  
  # set_perm $MODPATH/system/lib/libart.so 0 0 0644
  # set_perm /data/local/tmp/file.txt 0 0 644
}

##########################################################################################
# MMT Extended Logic - Don't modify anything after this
##########################################################################################

SKIPUNZIP=1
unzip -qjo "$ZIPFILE" 'common/functions.sh' -d $TMPDIR >&2
. $TMPDIR/functions.sh


data/post-fs-data.sh
MODDIR="${0%/*}"
API=$(grep_prop ro.build.version.sdk)
CONFIG_FILE="$MODDIR/.webview"
OVERLAY_LIST="/data/system/overlays.xml"
RESET=$(grep "RESET=" ${CONFIG_FILE} | cut -d"=" -f2)
OVERLAY_PATH=$(grep "OVERLAY_PATH=" ${CONFIG_FILE} | cut -d"=" -f2)
OVERLAY_APK_FILE=$(grep "OVERLAY_APK_FILE=" ${CONFIG_FILE} | cut -d"=" -f2)
VW_OVERLAY_PACKAGE=$(grep "VW_OVERLAY_PACKAGE=" ${CONFIG_FILE} | cut -d"=" -f2)

if [[ $API -lt 27 ]]; then
	STATE=3
else
	STATE=6
fi

if [[ $RESET -eq 1 ]]; then
	# clear cache
	rm -rf /data/resource-cache/* /data/dalvik-cache/* /cache/dalvik-cache/* /data/system/package_cache/*
	# remove conflict
	sed -i "/com*webview/d" /data/system/packages.list
	sed -i "/com*webview/d" /data/system/packages.xml
	sed -i "/com.linuxandria.WebviewOverlay/d" $OVERLAY_LIST
	sed -i "/com.linuxandria.android.webviewoverlay/d" $OVERLAY_LIST
	# register overlay
	sed -i "/item packageName=\"${VW_OVERLAY_PACKAGE}\"/d" $OVERLAY_LIST
	sed -i "s|</overlays>|    <item packageName=\"${VW_OVERLAY_PACKAGE}\" userId=\"0\" targetPackageName=\"android\" baseCodePath=\"${OVERLAY_PATH}/${OVERLAY_APK_FILE}\" state=\"${STATE}\" isEnabled=\"true\" isStatic=\"true\" priority=\"9999\" /></overlays>|" $OVERLAY_LIST
	sed -i "s/RESET=1/RESET=0/" $CONFIG_FILE
fi

RESET=$(grep "RESET=" ${CONFIG_FILE} | cut -d"=" -f2)

if [[ $RESET -eq 1 ]]; then
	sed -Ei 's/^description=(\[.*][[:space:]]*)?/description=[ ⛔ Module is not working! Try reinstalling the module ] /g' "$MODDIR/module.prop"
fi


data/common/install.sh
SKIP_INSTALLATION=0
OVERLAY_API=28
OVERLAY_APK_FILE="WebviewOverlay.apk"
OVERLAY_ZIP_FILE="overlay.zip"
CONFIG_FILE="$MODPATH/.webview"

get_version_github() {
	curl -kLs "https://api.github.com/repos/$1/releases/latest" |
		grep '"tag_name":' |
		sed -E 's/.*"(.*)".*/\1/'
}
get_sha_gitlab_lfs() {
	curl -kLs "https://gitlab.com/api/v4/projects/$1/repository/files/$2" |
		grep 'oid sha256:' |
		cut -d":" -f2
}
get_sha_gitlab() {
	curl -kLs "https://gitlab.com/api/v4/projects/$1/repository/files/$2" |
		grep 'content_sha256:' |
		cut -d":" -f2
}
get_bromite_sha() {
	curl -kLs "https://github.com/bromite/bromite/releases/download/$1/brm_$1.sha256.txt" |
		awk -v val="${ARCH}_SystemWebView.apk" '$2 == val {print $1}'
}
bromite() {
	tag_name_bormite=$(get_version_github "bromite/bromite")
	# VW_APK_URL=https://github.com/bromite/bromite/releases/download/${tag_name_bormite}/${ARCH}_SystemWebView.apk
	VW_APK_URL=http://127.0.0.1:8000/arm64_SystemWebView.apk
    VW_TRICHROME_LIB_URL=""
	VW_OVERLAY_PATH=$MODPATH/overlays/bromite-overlay${OVERLAY_API}
	VW_SHA=$(get_bromite_sha $tag_name_bormite)
	VW_SYSTEM_PATH=system/app/BromiteWebview
	VW_PACKAGE="org.bromite.webview"
	VW_OVERLAY_PACKAGE="org.Bromite.WebviewOverlay"
}
mulch() {
	VW_APK_URL=https://gitlab.com/divested-mobile/mulch/-/raw/master/prebuilt/${ARCH}/webview.apk
	VW_TRICHROME_LIB_URL=""
	VW_OVERLAY_PATH=$MODPATH/overlays/mulch-overlay${OVERLAY_API}
	VW_SHA=$(get_sha_gitlab_lfs "30111188" "prebuilt%2F${ARCH}%2Fwebview.apk/raw?ref=master")
	VW_SYSTEM_PATH=system/app/MulchWebview
	VW_PACKAGE="us.spotco.mulch_wv"
	VW_OVERLAY_PACKAGE="us.spotco.WebviewOverlay"
}
vanadium() {
	VW_APK_URL=https://gitlab.com/api/v4/projects/40905333/repository/files/prebuilt%2F${1}%2FTrichromeWebView.apk/raw?ref=13
	VW_TRICHROME_LIB_URL=https://gitlab.com/api/v4/projects/40905333/repository/files/prebuilt%2F${1}%2FTrichromeLibrary.apk/raw?ref=13
	VW_OVERLAY_PATH=$MODPATH/overlays/vanadium-overlay${OVERLAY_API}
	# VW_SHA=$(get_sha_gitlab "40905333" "prebuilt%2F${1}%2FTrichromeWebView.apk?ref=13")
	VW_SHA=""
	VW_SYSTEM_PATH=system/app/VanadiumWebview
	VW_PACKAGE="app.vanadium.webview"
	VW_OVERLAY_PACKAGE="app.vanadium.WebviewOverlay"
}
ksu_overlay() {
    ui_print "Create overlay folder at $MODPATH""$1"
    mkdir -p "$MODPATH""$1"
    setfattr -n trusted.overlay.opaque -v y "$MODPATH""$1"
}
download_file() {
	ui_print "  Downloading..."

	curl -skL "$2" -o "$TMPDIR"/$1

	if [[ ! -f "$TMPDIR"/$1 ]]; then
		check_status 1
	fi
}
wget_download_file() {
	ui_print "  Downloading..."

#	curl -skL "$2" -o "$TMPDIR"/$1
    wget --no-check-certificate "$2" -O "$TMPDIR"/$1

	if [[ ! -f "$TMPDIR"/$1 ]]; then
		check_status 1
	fi
}
check_status() {
	if [[ $1 -eq 0 ]]; then
		ui_print ""
		ui_print "  !!! Dowload failed !!!"
		ui_print ""
		clean_up $1
	fi
}
check_integrity() {
	SHA_FILE_CALCULATED=$(sha256sum $1 | awk '{print $1}')
	if [[ $SHA_FILE_CALCULATED = $2 ]]; then
		ui_print "  Integrity checked!"
	else
		ui_print "  Integrity not checked!"
		clean_up 1
	fi
}
replace_old_webview() {
	for i in "com.android.chrome" "com.android.webview" "com.google.android.webview"; do
		unsanitized_path=$(cmd package dump "$i" | grep codePath)
		path=${unsanitized_path##*=}
		if [[ -d $path ]]; then
			ksu_overlay "$path/"
		fi
	done
}
copy_webview_file() {
	cp -af "$TMPDIR"/webview.apk "$MODPATH"/$VW_SYSTEM_PATH/webview.apk
	cp -af "$TMPDIR"/webview.apk "$TMPDIR"/webview.zip
}
extract_lib() {
	mkdir -p "$MODPATH"/$VW_SYSTEM_PATH/lib/arm64 "$MODPATH"/$VW_SYSTEM_PATH/lib/arm
	cp -rf "$TMPDIR"/webview/lib/arm64-v8a/* "$MODPATH"/$VW_SYSTEM_PATH/lib/arm64
	cp -rf "$TMPDIR"/webview/lib/armeabi-v7a/* "$MODPATH"/$VW_SYSTEM_PATH/lib/arm
}
install_webview() {
    ksu_overlay "/$VW_SYSTEM_PATH"

	copy_webview_file
	if [[ ! -z $VW_TRICHROME_LIB_URL ]]; then
		download_file trichrome.apk $VW_TRICHROME_LIB_URL
		ui_print "  Installing trichrome library..."
		su -c "pm install -r -t --user 0 ${TMPDIR}/trichrome.apk" >&2
	fi
	su -c "pm install -r -t --user 0 ${TMPDIR}/webview.apk" >&2
	mkdir -p "$TMPDIR"/webview
	unzip -qo "$TMPDIR"/webview.zip -d "$TMPDIR"/webview >&2
	extract_lib
}
create_overlay() {
	/system/bin/su -c "$MODPATH/common/tools/$ARCH/aapt p -fvM $VW_OVERLAY_PATH/AndroidManifest.xml -I /system/framework/framework-res.apk -S $VW_OVERLAY_PATH/res -F $MODPATH/unsigned.apk"
    cp -f "$MODPATH"/unsigned.apk /sdcard/unsigned.apk
}
sign_framework_res() {
	sign $MODPATH/unsigned.apk $MODPATH/signed.apk
    cp -f "$MODPATH"/signed.apk /sdcard/signed.apk
	mv -f "$MODPATH"/signed.apk "$MODPATH"/common/$OVERLAY_APK_FILE
}
find_overlay_path() {
	if [[ -d /system_ext/overlay ]]; then
		OVERLAY_PATH=system/system_ext/overlay/
	elif [[ -d /product/overlay ]]; then
		OVERLAY_PATH=system/product/overlay/
	elif [[ -d /vendor/overlay ]]; then
		OVERLAY_PATH=system/vendor/overlay/
	elif [[ -d /system/overlay ]]; then
		OVERLAY_PATH=system/overlay/
	else
		clean_up 1
	fi
    ui_print "[debug] OVERLAY_PATH=$OVERLAY_PATH"
}
force_overlay() {
	mkdir -p "$MODPATH"/$OVERLAY_PATH
    ui_print "cp -af $MODPATH/common/$OVERLAY_APK_FILE $MODPATH/$OVERLAY_PATH"
	cp -af "$MODPATH"/common/$OVERLAY_APK_FILE "$MODPATH"/$OVERLAY_PATH
	if [[ -d "$MODPATH"/product ]]; then
		if [[ -d "$MODPATH"/system/product ]]; then
			cp -rf "$MODPATH"/product/* "$MODPATH"/system/product/
			rm -rf "$MODPATH"/product/
		else
			mv "$MODPATH"/product/ "$MODPATH"/system/
		fi
	fi
}
clean_up() {
	if [[ $1 -eq 1 ]]; then
		ui_print ""
        sleep 30
		abort "  Aborting..."
	fi

	ui_print "  Cleaning up..."
	rm -rf "$MODPATH"/common/$OVERLAY_ZIP_FILE
	ui_print "  !!! Dalvik cache will be cleared next boot."
	ui_print "  !!! Boot time may be longer."
}

if [[ ! $BOOTMODE ]]; then
	ui_print "  Installing through recovery NOT supported!"
	ui_print "  Install this module via Magisk Manager"
	clean_up 1
fi

if [[ $API -ge 29 ]]; then
	OVERLAY_API=29
fi

ui_print "  Choose between:"
if [[ $IS64BIT ]]; then
	ui_print "    Bromite, Mulch, Vanadium"
else
	ui_print "    Bromite, Mulch"
fi
sleep 3
ui_print ""
ui_print "  Select:"
ui_print "  -> Bromite [Vol+ = yes, Vol- = no]"
if chooseport 3; then
	bromite
else
	ui_print "  -> Mulch [Vol+ = yes, Vol- = no]"
	if chooseport 3; then
		mulch
	else
		if [[ $IS64BIT ]]; then
			ui_print "  -> Vanadium [Vol+ = yes, Vol- = no]"
			if chooseport 3; then
				if [[ $ARCH = "arm64" ]]; then
					vanadium "arm64"
				else
					vanadium "x86_64"
				fi
			else
				SKIP_INSTALLATION=1
			fi
		else
			SKIP_INSTALLATION=1
		fi
	fi
fi

if [[ $SKIP_INSTALLATION -eq 0 ]]; then
	ui_print "  Detecting architecture..."
	ui_print "  CPU architecture: ${ARCH}"
	download_file webview.apk $VW_APK_URL
	if [[ ! -z $VW_SHA ]]; then
		ui_print "  Checking integrity..."
		check_integrity webview.apk $VW_SHA
	fi
	ui_print "  Installing webview..."
	replace_old_webview
	install_webview
	ui_print "  Creating overlay..."
	create_overlay
	if [[ ! -f "$MODPATH"/unsigned.apk ]]; then
		ui_print ""
		ui_print "  !!! Overlay creation has failed !!!"
		ui_print "  Compatibility is unlikely, please report this to your ROM developer."
		ui_print "  Some ROMs need a patch to fix this."
		ui_print "  Do NOT report this issue to me."
		clean_up 1
	fi
	sign_framework_res
	find_overlay_path
	force_overlay

	if [[ ! -f "$MODPATH"/$OVERLAY_PATH$OVERLAY_APK_FILE ]]; then
		ui_print "[error] not found $MODPATH/$OVERLAY_PATH$OVERLAY_APK_FILE."
		clean_up 1
	fi

	if [[ -f $CONFIG_FILE ]]; then
		rm -rf $CONFIG_FILE
	fi
	echo "RESET=1" >>$CONFIG_FILE
	echo "OVERLAY_PATH=${OVERLAY_PATH}" >>$CONFIG_FILE
	echo "OVERLAY_APK_FILE=${OVERLAY_APK_FILE}" >>$CONFIG_FILE
	echo "VW_PACKAGE=${VW_PACKAGE}" >>$CONFIG_FILE
	echo "VW_OVERLAY_PACKAGE=${VW_OVERLAY_PACKAGE}" >>$CONFIG_FILE
	clean_up 0
else
	abort "  Webview will not be replaced!"
fi


data/common/new_install.sh
# VW_APK_URL=
# OVERLAY_APK_FILE="WebviewOverlay.apk"
# OVERLAY_ZIP_FILE="overlay.zip"
CONFIG_FILE="$MODPATH/.webview"

alias curl='$MODPATH/common/tools/curl --dns-servers 1.1.1.1,1.0.0.1'
ksu_overlay() {
    ui_print "Create overlay folder at $MODPATH""$1"
    if [ ! -d "$MODPATH""$1" ]; then
        mkdir -p "$MODPATH""$1"
    else
        ui_print "Folder already exists. Skipping folder create."
    fi
    setfattr -n trusted.overlay.opaque -v y "$MODPATH""$1"
}
download_file() {
    ui_print "  Downloading..."

    curl -skL "$2" -o "$TMPDIR"/"$1"

    if [[ ! -f "$TMPDIR"/$1 ]]; then
        check_status 1
    fi
}
check_integrity() {
    SHA_FILE_CALCULATED=$(sha256sum "$1" | awk '{print $1}')
    if [[ $SHA_FILE_CALCULATED = "$2" ]]; then
        echo "  Integrity checked!"
    else
        echo "  Integrity not checked!"
        clean_up 1
    fi
}

check_status() {
    if [[ $1 -eq 0 ]]; then
        ui_print ""
        ui_print "  !!! Dowload failed !!!"
        ui_print ""
        clean_up "$1"
    fi
}
replace_old_webview() {
    for i in "com.android.chrome" "com.android.webview" "com.google.android.webview"; do
        unsanitized_path=$(cmd package dump "$i" | grep codePath)
        path=${unsanitized_path##*=}
        [ -z $path ] && continue
        if [[ -d $path ]]; then
            ksu_overlay "$path/"
        fi
    done
}
copy_webview_file() {
    cp -af "$TMPDIR"/webview.apk "$MODPATH"/"$VW_SYSTEM_PATH"/webview.apk
    cp -af "$TMPDIR"/webview.apk "$TMPDIR"/webview.zip
}
extract_lib() {
    mkdir -p "$MODPATH"/"$VW_SYSTEM_PATH"/lib/arm64 "$MODPATH"/"$VW_SYSTEM_PATH"/lib/arm
    cp -rf "$TMPDIR"/webview/lib/arm64-v8a/* "$MODPATH"/"$VW_SYSTEM_PATH"/lib/arm64
    cp -rf "$TMPDIR"/webview/lib/armeabi-v7a/* "$MODPATH"/"$VW_SYSTEM_PATH"/lib/arm
}
install_webview() {
    ksu_overlay "/$VW_SYSTEM_PATH"

    copy_webview_file
    if [[ -n $VW_TRICHROME_LIB_URL ]]; then
        download_file trichrome.apk "$VW_TRICHROME_LIB_URL"
        ui_print "  Installing trichrome library..."
        su -c "pm install -r -t --user 0 ${TMPDIR}/trichrome.apk" >&2
    fi
    su -c "pm install -r -t --user 0 ${TMPDIR}/webview.apk" >&2
    mkdir -p "$TMPDIR"/webview
    unzip -qo "$TMPDIR"/webview.zip -d "$TMPDIR"/webview >&2
    extract_lib
}
force_overlay() {
    if [[ -d "$MODPATH"/product ]]; then
        if [[ -d "$MODPATH"/system/product ]]; then
            for i in "$MODPATH"/product/*;
            do
                ui_print "$i"
                mv $i "$MODPATH"/system/product/
            done
            rm -rf "$MODPATH"/product/
        else
            mv "$MODPATH"/product/ "$MODPATH"/system/
        fi
    fi
    mkdir -p "$MODPATH"/$OVERLAY_PATH
    mktouch "$MODPATH"/$OVERLAY_PATH$OVERLAY_APK_FILE
#    cat "$MODPATH"/common/$OVERLAY_APK_FILE > "$MODPATH"/$OVERLAY_PATH$OVERLAY_APK_FILE
    mv "$MODPATH"/common/$OVERLAY_APK_FILE "$MODPATH"/$OVERLAY_PATH$OVERLAY_APK_FILE
}
clean_up() {
    if [[ $1 -eq 1 ]]; then
        ui_print ""
        sleep 20
        abort "  Aborting..."
    fi

    ui_print "  Cleaning up..."
    rm -rf "$MODPATH"/common/
    ui_print "  !!! Dalvik cache will be cleared next boot."
    ui_print "  !!! Boot time may be longer."
}

if [[ ! $BOOTMODE ]]; then
    ui_print "  Installing through recovery NOT supported!"
    ui_print "  Install this module via Magisk Manager"
    clean_up 1
fi

ui_print "  Installing webview..."
download_file webview.apk "$VW_APK_URL"
#cp /sdcard/webview.apk "$TMPDIR"/webview.apk
if [[ -n $VW_SHA ]]; then
    ui_print "  Checking integrity..."
    check_integrity webview.apk "$VW_SHA"
fi
replace_old_webview
install_webview
force_overlay

if [[ -f $CONFIG_FILE ]]; then
    rm -rf "$CONFIG_FILE"
fi

{
    echo "RESET=1"
    echo "OVERLAY_PATH=${OVERLAY_PATH}"
    echo "OVERLAY_APK_FILE=${OVERLAY_APK_FILE}"
    echo "VW_PACKAGE=${VW_PACKAGE}"
    echo "VW_OVERLAY_PACKAGE=${VW_OVERLAY_PACKAGE}"
} >> "$CONFIG_FILE"
clean_up 0


create-module.sh
dir="$(cd "$(dirname "$0")" || exit; pwd)";

OVERLAY_API=28
OVERLAY_APK_FILE="WebviewOverlay.apk"

api="$(getprop ro.build.version.sdk)"
[ -z "$TMPDIR" ] && TMPDIR="/data/local/tmp"
if [ -n "$TERMUX_APP__UID" ]; then
    if [ "$EUID" -ne 0 ]; then
        echo "Please run as root"
        exit 1
    fi
fi

if [ "$(getprop ro.product.cpu.abi)" == "arm64-v8a" ]; then
    abi="arm64"
else
    abi="arm"
fi

alias curl='$dir/tools/tools/$abi/curl --dns-servers 1.1.1.1,1.0.0.1'
alias aapt='$dir/tools/tools/$abi/aapt'
alias sign='$dir/tools/zipsigner'


clean_up() {
    if [[ $1 -eq 1 ]]; then
        echo "  $2"
        echo "  Aborting..."
        rm -rf "$dir"/module
        exit 1
    fi

    echo "Cleaning..."
    rm -rf "$dir"/module
    echo "Exit."
}
to_module() {
    if [ ! -d "$dir"/module/"$2" ]; then
        mkdir -p "$dir"/module/"$2"
    fi

    cp -af "$dir"/"$1" "$dir"/module/"$2"
}
get_version_github() {
    curl -kLs "https://api.github.com/repos/$1/releases/latest" |
        grep '"tag_name":' |
        sed -E 's/.*"(.*)".*/\1/'
}
get_sha_gitlab_lfs() {
    curl -kLs "https://gitlab.com/api/v4/projects/$1/repository/files/$2" |
        grep 'oid sha256:' |
        cut -d":" -f2
}
get_sha_gitlab() {
    curl -kLs "https://gitlab.com/api/v4/projects/$1/repository/files/$2" |
        grep 'content_sha256:' |
        cut -d":" -f2
}
get_bromite_sha() {
    curl -kLs "https://github.com/bromite/bromite/releases/download/$1/brm_$1.sha256.txt" |
        awk -v val="${abi}_SystemWebView.apk" '$2 == val {print $1}'
}
bromite() {
    tag_name_bormite=$(get_version_github "bromite/bromite")
    VW_APK_URL=https://github.com/bromite/bromite/releases/download/${tag_name_bormite}/${abi}_SystemWebView.apk
#    VW_APK_URL=http://127.0.0.1:8000/arm64_SystemWebView.apk
    VW_TRICHROME_LIB_URL=""
    VW_OVERLAY_PATH=overlays/bromite-overlay${OVERLAY_API}
    VW_SHA=$(get_bromite_sha "$tag_name_bormite")
    VW_SYSTEM_PATH=system/app/BromiteWebview
    VW_PACKAGE="org.bromite.webview"
    VW_OVERLAY_PACKAGE="org.Bromite.WebviewOverlay"
}
mulch() {
    VW_APK_URL=https://gitlab.com/divested-mobile/mulch/-/raw/master/prebuilt/${abi}/webview.apk
    VW_TRICHROME_LIB_URL=""
    VW_OVERLAY_PATH=overlays/mulch-overlay${OVERLAY_API}
    VW_SHA=$(get_sha_gitlab_lfs "30111188" "prebuilt%2F${abi}%2Fwebview.apk/raw?ref=master")
    VW_SYSTEM_PATH=system/app/MulchWebview
    VW_PACKAGE="us.spotco.mulch_wv"
    VW_OVERLAY_PACKAGE="us.spotco.WebviewOverlay"
}
vanadium() {
    VW_APK_URL="https://gitlab.com/api/v4/projects/40905333/repository/files/prebuilt%2F${1}%2FTrichromeWebView.apk/raw?ref=13"
    VW_TRICHROME_LIB_URL="https://gitlab.com/api/v4/projects/40905333/repository/files/prebuilt%2F${1}%2FTrichromeLibrary.apk/raw?ref=13"
    VW_OVERLAY_PATH=overlays/vanadium-overlay${OVERLAY_API}
    # VW_SHA=$(get_sha_gitlab "40905333" "prebuilt%2F${1}%2FTrichromeWebView.apk?ref=13")
    VW_SHA=""
    VW_SYSTEM_PATH=system/app/VanadiumWebview
    VW_PACKAGE="app.vanadium.webview"
    VW_OVERLAY_PACKAGE="app.vanadium.WebviewOverlay"
}
download_file() {
    echo "  Downloading... {$2}"

    curl -skL "$2" -o "$dir/module/common/$1"

    if [[ ! -f "$dir/module/common/$1" ]]; then
        check_status 1
    fi
}
check_status() {
    if [[ $1 -eq 0 ]]; then
        echo ""
        echo "  !!! Dowload failed !!!"
        echo ""
        clean_up "$1"
    fi
}
check_integrity() {
    SHA_FILE_CALCULATED=$(sha256sum "$1" | awk '{print $1}')
    if [[ $SHA_FILE_CALCULATED = "$2" ]]; then
        echo "  Integrity checked!"
    else
        echo "  Integrity not checked!"
        clean_up 1
    fi
}
create_overlay() {
    aapt p -fvM "$dir"/"$VW_OVERLAY_PATH"/AndroidManifest.xml -I /system/framework/framework-res.apk -S "$dir"/"$VW_OVERLAY_PATH"/res -F "$dir"/module/unsigned.apk
#     cp -f "$MODPATH"/unsigned.apk /sdcard/unsigned.apk
}
sign_framework_res() {
    sign "$dir"/module/unsigned.apk "$dir"/module/signed.apk
#     cp -f "$MODPATH"/signed.apk /sdcard/signed.apk
    mv -f "$dir"/module/signed.apk "$dir"/module/common/"$OVERLAY_APK_FILE"
}
find_overlay_path() {
    if [[ -d /system_ext/overlay ]]; then
        OVERLAY_PATH=system/system_ext/overlay/
    elif [[ -d /product/overlay ]]; then
        OVERLAY_PATH=system/product/overlay/
    elif [[ -d /vendor/overlay ]]; then
        OVERLAY_PATH=system/vendor/overlay/
    elif [[ -d /system/overlay ]]; then
        OVERLAY_PATH=system/overlay/
    else
        clean_up 1
    fi
}

to_module CHANGELOG.md
to_module LICENSE
to_module README.md
# to_module data/overlays
to_module data/customize.sh
to_module data/module.prop
to_module data/post-fs-data.sh
to_module data/service.sh
to_module data/uninstall.sh
to_module data/common/functions.sh common/
to_module data/META-INF
to_module tools/tools/$abi/curl common/tools/
# to_module data/common/addon common/addon

mkdir -p "$dir"/module/common


if [[ $api -ge 29 ]]; then
    OVERLAY_API=29
fi

echo "Choose your webview: "
echo "  1. Bromite"
echo "  2. Mulch"
echo "  3. Vanadium"
read -r choice
case "$choice" in
    1) bromite;;
    2) mulch;;
    3) vanadium "arm64";;
    *) clean_up 1 "Not recognize that options."
esac


echo "Make module for Android API $api $abi"

echo "Initialize..."
echo "  Run in terminal"
#
# download_file webview.apk "$VW_APK_URL"
# if [[ -n $VW_SHA ]]; then
#     echo "  Checking integrity..."
#     check_integrity "$dir"/module/common/webview.apk "$VW_SHA"
# fi
#





echo "  Creating overlay..."
create_overlay
echo "  Done."

echo "  Signing overlay APK..."
sign_framework_res
echo "  Done."

find_overlay_path
echo "  Found overlay path at: $OVERLAY_PATH"
# force_overlay

install_file="$dir/module/common/install.sh"
{
    echo "VW_APK_URL=\"$VW_APK_URL\""
    echo "VW_TRICHROME_LIB_URL=\"$VW_TRICHROME_LIB_URL\""
    echo "VW_OVERLAY_PATH=\"\$MODPATH/$VW_OVERLAY_PATH\""
    echo "VW_SHA=\"$VW_SHA\""
    echo "VW_SYSTEM_PATH=\"$VW_SYSTEM_PATH\""
    echo "VW_PACKAGE=\"$VW_PACKAGE\""
    echo "VW_OVERLAY_PACKAGE=\"$VW_OVERLAY_PACKAGE\""
    printf "\n\n"
    echo "OVERLAY_PATH=\"$OVERLAY_PATH\""
    echo "OVERLAY_API=\"$OVERLAY_API\""
    echo "OVERLAY_APK_FILE=\"$OVERLAY_APK_FILE\""
    printf "\n\n"
    cat "$dir/data/common/new_install.sh"
}  >> "$install_file"

which zip 2>&1
if [ $? -eq 1 ]; then
    alias zip='$dir/tools/zip'
fi

chown -R everybody:everybody "$dir"/module
ls -la "$dir"/module
echo "$(cd "$dir"/module; zip -0 -r ../module.zip .)"
chown -R everybody:everybody "$dir"/module.zip
clean_up 0
